### PR TITLE
First integration tools

### PR DESCRIPTION
**Pull request by gparant**
# Integrate: 
## In the map editor (area & object):
 - Google doc
 - Google Excel
 - Google Slide
 - Youtube
 - Klaxoon
 
## In the app menu:
 - Klaxoon Activity Picker
 -  Google doc
 - Google Excel
 - Google Slide
 
 ## In the chat
 - Google doc
 - Google Excel
 - Google Slide
 - Youtube
 - Klaxoon

### CODE REVIEW

File: chat/src/Components/ChatMessageForm.svelte [hunk 1]
```diff
@@ -321,6 +529,57 @@
...
+                    {/if}
+                    {#if app.name === "Google Docs"}
+                        {$LL.form.application.googleDocs.description()}
+                    {/if}
+                    {#if app.name === "Google Sheets"}
+                        {$LL.form.application.googleSheets.description()}
+                    {/if}
+                    {#if app.name === "Google Slides"}
+                        {$LL.form.application.googleSlides.description()}
+                    {/if}
```
Comment (moufmouf) [inline chat/src/Components/ChatMessageForm.svelte:553]: This looks suspiciously hard to maintain. Maybe the "app" object could contain a pointer to the `$LL.form.application[app_key]` translation?

File: chat/src/Components/ChatMessageForm.svelte [hunk 2]
```diff
@@ -46,6 +56,17 @@
     let newMessageText = "";
     let htmlMessageText = "";
     let usersSearching: User[] = [];
+    let applicationMenuIsOpenned = false;
+
+    interface Application {
+        name: string;
+        icon: string;
+        exemple: string;
```
Comment (moufmouf) [inline chat/src/Components/ChatMessageForm.svelte:64]: That's "example" in English :)

File: libs/shared-utils/src/Application/YoutubeService.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    return await axios
+        .get(`https://www.youtube.com/oembed?url=${link}&format:json`)
+        .then((res: AxiosResponse<{ html: string | undefined }>) => {
+            const html = res.data.html;
+            if (html == undefined) throw new Error("No html found");
+            const div = document.createElement("div");
+            div.insertAdjacentHTML("beforeend", html);
+            const iframe: HTMLIFrameElement = div.firstChild as HTMLIFrameElement;
+            return iframe.src;
+        })
```
Comment (moufmouf) [inline libs/shared-utils/src/Application/YoutubeService.ts:16]: This piece of code is generic to the oEmbed protocol and not specific to Youtube. We could maybe put it in a more generic service.

File: play/public/resources/help-popup-blocked/1-popupblocked.png
Comment (moufmouf) [inline play/public/resources/help-popup-blocked/1-popupblocked.png:1]: You should the screenshot to "english"

Comment (moufmouf) [inline play/public/resources/help-popup-blocked/1-popupblocked.png:1]: You should the screenshot to "english"

**Review by moufmouf — COMMENTED** [review]
Pretty cool, but we could make this even more generic. Here is a proposal:

We already have an endpoint in the pusher that is checking if a page can be embedded in an iframe. It is performing a request on the URL and looking at the headers (here: https://github.com/thecodingmachine/workadventure/blob/6451196d2f6ac11eff8890f0ee51b5241c64402a/play/src/pusher/services/SocketManager.ts#L1187-L1187)

My idea is to extend this function. In addition to the checking for the `x-frame-options` header, we could load the page and look at the OEmbed URL in a "<link>" tag (as specified in the Discovery chapter of the OEmbed spec: https://oembed.com/#section4)

From here, we can get the OEmbed data, parse the HTML and finally extract the URL of the iframe.

Alternatively, we can use a third party container like "IFramely": https://github.com/itteco/iframely
Also, we could use this request to extract the favicon URL from the web page.

Ideally, instead of replacing the URL when it is pasted in the input field, I would call the service at the last moment. For instance, when someone clicks on a link (in the chat). If the website is embeddable (or iframable), we could display a small popup asking where to open the link (new tab or embedded). Even better, in the future, we could propose to open the website for everybody automatically (especially in bubbles or if you are a speaker)

Lastly, some URLs may not support oEmbed and may require special code (like for Google Workspace URLs)

Here, we could make the system more generic by extending the data returned by the Admin API.

Instead of:

```
{
   "youtubeToolActivated": true,
   "googleDocsToolActivated": true,
   "googleSheetsToolActivated": true,
}
```
 we could send something like:
 
 
```json
{
    "thirdParty": {
        "apps": [
            {
                "name": "Youtube"
                "label": "Embed Youtube video",
                "iconUrl": "...",
                "editorComment": "Use this link to embed....",
                "oEmbedUrl": "https://youtube.com/oEmbed",
            },
            {
                "name": "Google Slides"
                "label": "Embed Google Slides presentation",
                "editorComment": "Use this link to embed....",
                "iconUrl": "...",
                "urlRegex": "https://docs.google.com/slides/(.*)",
                "urlReplace": "https://docs.google.com/slides/$1?embeddable=true",
            },
        ]
    }
}
```

Doing it this way could be awesome because it would be tremendously easy to add new tools (think Office 365...)

What do you think?

File: docs/schema/1.0/wam.json [hunk 4]
```diff
@@ -168,6 +171,25 @@
...
+                        "application": {
+                          "type": "string",
+                          "enum": [
+                            "website",
+                            "youtube",
+                            "klaxoon",
+                            "googleDocs",
+                            "googleSheets",
+                            "googleSlides",
+                            "googleForms"
```
[the remark this answers is not in the document]
Comment (gparant) [reply docs/schema/1.0/wam.json:184]: The default of application is `website`, this enum is only to manage correctly object into map storage and chat

File: libs/shared-utils/src/Application/YoutubeService.ts (hunk 3 above)
Comment (gparant) [reply libs/shared-utils/src/Application/YoutubeService.ts:16]: if we integrate new service with `oembed` protocol, we could create a generic service and use it to our applications

**Comment by gparant** [comment]
> 

Really interesting. Today we already use third party to check embedded link: https://embed.ly/

Today we use `queryEmbeddableWebsite` function created for map storage to check if one link is embed or not: https://github.com/workadventure/workadventure/blob/2f27ea81cffcd24ace8cf70f5c30a58efd8fb5a8/play/src/front/Connection/RoomConnection.ts#L1481

If we integrate new service with a oEmbed protocol, we could create new generic service to get embedded link and check if it could be opened with a website interface.

**Comment by gparant** [comment]
> Pretty cool, but we could make this even more generic. Here is a proposal:
> 
> We already have an endpoint in the pusher that is checking if a page can be embedded in an iframe. It is performing a request on the URL and looking at the headers (here: https://github.com/thecodingmachine/workadventure/blob/6451196d2f6ac11eff8890f0ee51b5241c64402a/play/src/pusher/services/SocketManager.ts#L1187-L1187)
> 
> My idea is to extend this function. In addition to the checking for the `x-frame-options` header, we could load the page and look at the OEmbed URL in a "" tag (as specified in the Discovery chapter of the OEmbed spec: https://oembed.com/#section4)
> 
> From here, we can get the OEmbed data, parse the HTML and finally extract the URL of the iframe.
> 
> Alternatively, we can use a third party container like "IFramely": https://github.com/itteco/iframely Also, we could use this request to extract the favicon URL from the web page.
> 
> Ideally, instead of replacing the URL when it is pasted in the input field, I would call the service at the last moment. For instance, when someone clicks on a link (in the chat). If the website is embeddable (or iframable), we could display a small popup asking where to open the link (new tab or embedded). Even better, in the future, we could propose to open the website for everybody automatically (especially in bubbles or if you are a speaker)
> 
> Lastly, some URLs may not support oEmbed and may require special code (like for Google Workspace URLs)
> 
> Here, we could make the system more generic by extending the data returned by the Admin API.
> 
> Instead of:
> 
> ```
> {
>    "youtubeToolActivated": true,
>    "googleDocsToolActivated": true,
>    "googleSheetsToolActivated": true,
> }
> ```
> 
> we could send something like:
> 
> ```json
> {
>     "thirdParty": {
>         "apps": [
>             {
>                 "name": "Youtube"
>                 "label": "Embed Youtube video",
>                 "iconUrl": "...",
>                 "editorComment": "Use this link to embed....",
>                 "oEmbedUrl": "https://youtube.com/oEmbed",
>             },
>             {
>                 "name": "Google Slides"
>                 "label": "Embed Google Slides presentation",
>                 "editorComment": "Use this link to embed....",
>                 "iconUrl": "...",
>                 "urlRegex": "https://docs.google.com/slides/(.*)",
>                 "urlReplace": "https://docs.google.com/slides/$1?embeddable=true",
>             },
>         ]
>     }
> }
> ```
> 
> Doing it this way could be awesome because it would be tremendously easy to add new tools (think Office 365...)
> 
> What do you think?

Really interesting. Today we already use third party to check embedded link: https://embed.ly/

Today we use queryEmbeddableWebsite function created for map storage to check if one link is embed or not:
https://github.com/workadventure/workadventure/blob/2f27ea81cffcd24ace8cf70f5c30a58efd8fb5a8/play/src/front/Connection/RoomConnection.ts#L1481C18-L1492

If we integrate new service with a oEmbed protocol, we could create new generic service to get embedded link and check if it could be opened with a website interface.

File: tests/tests/chat.spec.ts [hunk 5]
```diff
@@ -196,6 +196,69 @@ test.describe('Chat', () => {
...
+    if(browserName === "webkit"){
+      // Because webkit in playwright does not support Camera/Microphone Permission by settings
+      await hideNoCamera(page);
+    }
+
+    await Menu.openChat(page);
+    await Chat.openTimeline(page);
+
+    // Create test to use Docs application
+    await page.frameLocator('iframe#chatWorkAdventure').locator('aside.chatWindow').locator('.actions').first().click();
```
Comment (moufmouf) [inline tests/tests/chat.spec.ts:217]: We could have a 

```
const chatWindow = page.frameLocator('iframe#chatWorkAdventure').locator('aside.chatWindow');
```

That would be easier to read.

File: tests/tests/iframe_script.spec.ts [hunk 6]
```diff
@@ -16,17 +16,20 @@ test.describe('Iframe API', () => {
...
     );
 
     await login(page);
 
-    await page.click('#menuIcon img:first-child');
-
-    await page.click('button:has-text("custom iframe menu")');
+    // wait some time for the scripting menu to be loaded
+    // eslint-disable-next-line playwright/no-wait-for-timeout
+    await page.waitForTimeout(2_000);
```
Comment (moufmouf) [inline tests/tests/iframe_script.spec.ts:29]: Ach! Verbotten!

I'm doing my best to keep timeouts out of the tests. They are too slow already. We need to replace that with a proper "expect".

File: libs/shared-utils/.eslintrc.js [hunk 7]
```diff
@@ -30,6 +30,7 @@ module.exports = {
       "error", { "args": "none", "caughtErrors": "all", "varsIgnorePattern": "_exhaustiveCheck" }
     ],
     "import/order": "error",
+    "@typescript-eslint/ban-ts-comment": "off",
```
[the remark this answers is not in the document]
Comment (gparant) [reply libs/shared-utils/.eslintrc.js:33]: Unfortunely I need skip some check for KlaxoonActivityPicker

**Pull request merged**